sys/kern/interrupt.c

```diff
@@ -127,60 +141,80 @@ void intr_root_handler(ctx_t *ctx) {
     on_user_exc_leave();
 }
 
-/* interrupt handlers delegated to be called in the interrupt thread */
-static ih_list_t delegated = TAILQ_HEAD_INITIALIZER(delegated);
+static void intr_thread_create(intr_event_t *ie) {
+  intr_thread_t *it = kmalloc(M_INTR, sizeof(intr_thread_t), M_ZERO);
+
+  ie->ie_ithread = it;
+  it->it_event = ie;
+  it->it_thread = thread_create(ie->ie_name, intr_thread, it, prio_ithread(0));
+
+  sched_add(it->it_thread);
+}
 
 void intr_event_run_handlers(intr_event_t *ie) {
   intr_handler_t *ih, *next;
-  intr_filter_t status = IF_STRAY;
 
+  assert(intr_disabled());
   assert(ie != NULL);
 
+  /* Do we wake up an ithread */
+  intr_filter_t ie_status = IF_STRAY;
+
   TAILQ_FOREACH_SAFE (ih, &ie->ie_handlers, ih_link, next) {
-    status = ih->ih_filter(ih->ih_argument);
-    if (status == IF_FILTERED)
-      return;
+    intr_filter_t status = ih->ih_filter(ih->ih_argument);
+
+    ie_status |= status;
+
     if (status == IF_DELEGATE) {
       assert(ih->ih_service);
+      ih->ih_flags = IH_DELEGATE;
+    }
+  }
 
-      if (ie->ie_disable)
-        ie->ie_disable(ie);
+  if (ie_status & IF_DELEGATE) {
+    if (ie->ie_disable)
+      ie->ie_disable(ie);
 
-      TAILQ_REMOVE(&ie->ie_handlers, ih, ih_link);
-      TAILQ_INSERT_TAIL(&delegated, ih, ih_link);
-      sleepq_signal(&delegated);
-      return;
-    }
+    if (ie->ie_ithread == NULL)
+      intr_thread_create(ie);
```

Comment:
`intr_thread_create` calls `kmalloc`, which we **must not** do in the bottom-half. I'd suggest calling `kmalloc` with `M_NOWAIT` flag, but it looks like it'd try to acquire a mutex anyway, so it isn't allowed either.

What about:
* creating the ithread (or a pool of them) somewhere else, earlier, in the top-half?
* or if you want to create the ithread here, consider using some pre-allocated static memory (but please note that `thread_create` also internally uses mutexes)

Reply:
Oops... thank you for spotting that. Indeed this is wrong place to create a thread. I'll move it to `ie_add_handler` most likely.